File: Java-Core/src/main/java/presto/value/ExpressionValue.java
package presto.value;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.type.IType;
import presto.utils.CodeWriter;

/* a value which can be used as an expression in a check/interpret scenario */
public class ExpressionValue extends BaseValue implements IExpression {

	IValue value;
	
	public ExpressionValue(IType type, IValue value) {
		super(type);
		this.value = value;
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		return type;
	}

	@Override
	public IValue interpret(Context context) throws PrestoError {
		return value;
	}
	
	@Override
	public String toString() {
		return value.toString();
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		writer.append(value.toString()); // value has no dialect
	}
	
	@Override
	public ISliceable<IValue> asSliceable(Context context) throws PrestoError {
		return value.asSliceable(context);
	}
	

}

File: Java-CodeServer/src/main/resources/js/prompto/value/ExpressionValue.js
var Value = require("./Value").Value;

function ExpressionValue(type, value) {
	Value.call(this, type);
	this.value = value;
    // make this sliceable
    this.sliceable = value.slice ? value : null;
	return this;
}

ExpressionValue.prototype = Object.create(Value.prototype);
ExpressionValue.prototype.constructor = ExpressionValue;

ExpressionValue.prototype.check = function(context) {
	return this.type;
};

ExpressionValue.prototype.interpret = function(context) {
	if(this.value.interpret) {
		return this.value.interpret(context);
	} else {
		return this.value;
	}
};

ExpressionValue.prototype.toString = function() {
	return this.value.toString();
};

ExpressionValue.prototype.toDialect = function(dialect) {
	return this.value.toDialect(dialect);
};

exports.ExpressionValue = ExpressionValue;
